analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs

```diff
@@ -428,3 +514,113 @@ public void NestedCombined(DateTime? value)
         }
     }
 }
+
+class Casts
+{
+    void Downcast1(int? i)
+    {
+        _ = (int)i;                   // Compliant, same as i.Value with i unknown
+        i = null;
+        _ = (int)i;                   // Noncompliant, empty
+        i = 42;
+        _ = (int)i;                   // Compliant, unreachable
+    }
+
+    void Downcast2(int? i)
+    {
+        i = 42;
+        _ = (int)i;                   // Compliant, non-empty
+    }
+
+    void Downcast3()
+    {
+        int? i = null;
+        _ = i.Value;                  // Noncompliant, empty
+        _ = (int)i;                   // Compliant, unreachable
+    }
+
+    void Upcast1(int? i1, int? i2)
+    {
+        _ = ((int?)null).Value;       // Noncompliant
+    }
+
+    void Upcast2(int? i1, int? i2)
+    {
+        _ = ((int?)null).Value;       // Noncompliant
+        i2 = null;
+        _ = ((int?)i2).Value;         // Noncompliant, FP: unreachable
```

Comment:
Line 551 is truly unreachable: line 549 should learn that null is not null, because `.Value` is called on it, and there should not be program state where such contradiction should exist.
Therefore, there shouldn't be a path to line 551, and no issue should be raised there.

Reply:
This is still not FP to me. The engine should not stop exploring so we don't want to fix this by-design.

At that line, for whatever reason we reach it (like engine imperfections), we know that `i` is null. So `i.Value` should raise an issue. The `FP` comment is confusing.

Reply:
It's same case as line 556 below